### PR TITLE
Implement `cudf::reduce` for `decimal32` and `decimal64` (part 1)

### PR DESCRIPTION
**Pull request by codereport**
This PR resolves a part of https://github.com/rapidsai/cudf/issues/3556.

Might split this up into multiple PRs. Will focus on `MIN`, `MAX`, `SUM` & `PRODUCT` for now.

**Reduction Ops:**
```
  enum Kind {
    SUM,             ///< sum reduction
    PRODUCT,         ///< product reduction
    MIN,             ///< min reduction
    MAX,             ///< max reduction
    COUNT_VALID,     ///< count number of valid elements
    COUNT_ALL,       ///< count number of elements
    ANY,             ///< any reduction
    ALL,             ///< all reduction
    SUM_OF_SQUARES,  ///< sum of squares reduction
    MEAN,            ///< arithmetic mean reduction
    VARIANCE,        ///< groupwise variance
    STD,             ///< groupwise standard deviation
    MEDIAN,          ///< median reduction
    QUANTILE,        ///< compute specified quantile(s)
    ARGMAX,          ///< Index of max element
    ARGMIN,          ///< Index of min element
    NUNIQUE,         ///< count number of unique elements
    NTH_ELEMENT,     ///< get the nth element
    ROW_NUMBER,      ///< get row-number of element
    COLLECT,         ///< collect values into a list
    LEAD,            ///< window function, accesses row at specified offset following current row
    LAG,             ///< window function, accesses row at specified offset preceding current row
    PTX,             ///< PTX UDF based reduction
    CUDA             ///< CUDA UDf based reduction
  };
```
**To Do List:**
* [x] `SUM`
   * [x] Implementation
   * [x] Basic unit tests
   * [x] Comprehensive unit tests
* [x] `PRODUCT`
   * [x] Implementation
   * [x] Basic unit tests
   * [x] Comprehensive unit tests
* [x] `MAX`
   * [x] Implementation
   * [x] Basic unit tests
   * [x] Comprehensive unit tests
* [x] `MIN`
   * [x] Implementation
   * [x] Basic unit tests
   * [x] Comprehensive unit tests
#### Operations that "fell out":
* [x] `NUNIQUE`
   * [x] Implementation
   * [x] Basic unit tests

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

File: cpp/src/reductions/reductions.cpp [hunk 1]
```diff
@@ -88,19 +88,19 @@ struct reduce_dispatch_functor {
...
+      }
       case aggregation::NUNIQUE: {
         auto nunique_agg = static_cast<nunique_aggregation const *>(agg.get());
         return make_fixed_width_scalar(
           detail::distinct_count(
             col, nunique_agg->_null_handling, nan_policy::NAN_IS_VALID, stream.value()),
           stream.value(),
           mr);
-      } break;
+      }
```
Comment (karthikeyann) [inline cpp/src/reductions/reductions.cpp:99]: Isn't compiler throwing a warning here for leaving out `break;`?

Comment (codereport) [reply cpp/src/reductions/reductions.cpp:99]: I don't think so? all of the cases have a `return` - which makes `break` unnecessary (IIUC).

**Review by rgsl888prabhu — APPROVED** [review]
LGTM, small questions and changes.

File: cpp/include/cudf/detail/reduction.cuh [hunk 2]
```diff
@@ -99,9 +116,9 @@ std::unique_ptr<scalar> reduce(InputIterator d_in,
                                rmm::cuda_stream_view stream,
                                rmm::mr::device_memory_resource* mr)
 {
-  auto binary_op      = sop.get_binary_op();
-  OutputType identity = sop.template get_identity<OutputType>();
-  rmm::device_scalar<OutputType> dev_result{identity, stream};
+  auto binary_op  = sop.get_binary_op();
+  auto identity   = sop.template get_identity<OutputType>();
+  auto dev_result = rmm::device_scalar<OutputType>{identity, stream, mr};
```
Comment (karthikeyann) [inline cpp/include/cudf/detail/reduction.cuh:121]: from `string_view`, it's not moved. It's copied, hence the missing `mr`.

**Review by karthikeyann — CHANGES_REQUESTED** [review]

File: cpp/src/reductions/simple.cuh [hunk 3]
```diff
@@ -28,6 +28,7 @@
 #include <cudf/utilities/type_dispatcher.hpp>
 
 #include <rmm/cuda_stream_view.hpp>
+#include "cudf/utilities/error.hpp"
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/reductions/simple.cuh]: Fixed.

File: cpp/src/reductions/simple.cuh [hunk 4]
```diff
@@ -50,26 +51,58 @@ std::unique_ptr<scalar> simple_reduction(column_view const& col,
...
+      auto it = thrust::make_transform_iterator(dcol->pair_begin<Type, true>(), f);
+      return detail::reduce(it, col.size(), simple_op, stream, mr);
+    } else {
+      auto f  = simple_op.template get_element_transformer<ResType>();
+      auto it = thrust::make_transform_iterator(dcol->begin<Type>(), f);
+      return detail::reduce(it, col.size(), simple_op, stream, mr);
+    }
+  }();
+
+  // set scalar is valid0
```
[the remark this answers is not in the document]
Comment (codereport) [reply cpp/src/reductions/simple.cuh]: Good catch

File: cpp/include/cudf/scalar/scalar.hpp [hunk 5]
```diff
@@ -377,6 +391,7 @@ class fixed_point_scalar : public scalar {
 
  protected:
   rmm::device_scalar<rep_type> _data{};  ///< device memory containing the value
+  rmm::device_scalar<int32_t> _scale{};  ///< device memory containing the scale
```
Comment (davidwendt) [inline cpp/include/cudf/scalar/scalar.hpp:394]: Seems we are storing the scale in this class twice. Once in the base class's `_data_type` and once here. I just worry they could get out synch. Perhaps just create an accessor in this class that reads it out of the protected `_data_type` member?

**Review by davidwendt — CHANGES_REQUESTED** [review]

File: cpp/include/cudf/scalar/scalar.hpp (hunk 5 above)
Comment (codereport) [reply cpp/include/cudf/scalar/scalar.hpp:394]: Yea, honestly I have no I idea what I was thinking here at the time. I should have been just accessing the value in the base `scalar` class. Fix incoming.

**Review by davidwendt — APPROVED** [review]

**Review by karthikeyann — APPROVED** [review]

**Pull request merged**